internal/config/command_line.go

```diff
@@ -0,0 +1,142 @@
+package config
+
+import (
+	"flag"
+	"io"
+	"os"
+	"slices"
+	"strconv"
+	"strings"
+
+	"github.com/yunginnanet/HellPot/internal/extra"
+	"github.com/yunginnanet/HellPot/internal/version"
+)
+
+var CLIFlags = flag.NewFlagSet("config", flag.ContinueOnError)
+
+var (
+	sliceDefs = make(map[string][]string)
+	slicePtrs = make(map[string]*string)
+)
+
+func addCLIFlags() {
+	parse := func(k string, v interface{}, nestedName string) {
+		switch casted := v.(type) {
+		case bool:
+			CLIFlags.Bool(nestedName, casted, "set "+k)
+		case string:
+			CLIFlags.String(nestedName, casted, "set "+k)
+		case int:
+			CLIFlags.Int(nestedName, casted, "set "+k)
+		case float64:
+			CLIFlags.Float64(nestedName, casted, "set "+k)
+		case []string:
+			sliceDefs[nestedName] = casted
+			joined := strings.Join(sliceDefs[nestedName], ",")
+			slicePtrs[nestedName] = CLIFlags.String(nestedName, joined, "set "+k)
+		}
+	}
+
+	for key, val := range Defaults.val {
+		if _, ok := val.(map[string]interface{}); !ok {
+			parse(key, val, key)
+			continue
+		}
+		nested, ok := val.(map[string]interface{})
+		if !ok {
+			// linter was confused by the above check
+			panic("unreachable, if you see this you have entered a real life HellPot")
+		}
+		for k, v := range nested {
+			nestedName := key + "." + k
+			parse(k, v, nestedName)
+		}
+	}
+}
+
+var replacer = map[string][]string{
+	"-h": {"-help"},
+	"-v": {"-version"},
+	"-c": {"-config"},
+	"-g": {"-bespoke.enable_grimoire", "true", "-bespoke.grimoire_file"},
+}
+
+func InitCLI() {
+	newArgs := make([]string, 0)
+	for _, arg := range os.Args {
+		if repl, ok := replacer[arg]; ok {
+			newArgs = append(newArgs, repl...)
+			continue
+		}
+		// check for unit test flags
+		if !strings.HasPrefix(arg, "-test.") {
+			newArgs = append(newArgs, arg)
+		}
+	}
+
+	newArgs = slices.Compact(newArgs)
+
+	CLIFlags.Bool("banner", false, "show banner and version then exit")
+	CLIFlags.Bool("genconfig", false, "write default config to stdout then exit")
+	CLIFlags.Bool("help", false, "show this help and exit")
+	CLIFlags.String("config", "", "specify config file")
+	CLIFlags.String("version", "", "show version and exit")
+
+	addCLIFlags()
+
+	if err := CLIFlags.Parse(newArgs[1:]); err != nil {
+		println(err.Error())
+		os.Exit(2)
+	}
+
+	for defaultKey, defaultVal := range Defaults.val {
```

Comment:
Noting (for myself even): I was doing this here because of some problematic behavior when it comes to order of initialization and overrides when it comes to CLI vs defaults vs environment variables